File: Dynamic_Programming/1751. Maximum Number of Events That Can Be Attended II.cpp
// TC:O(n * k + n * lgn) SC:O(n * k)
class Solution {
    using ll = long long;
public:
    int maxValue(vector<vector<int>>& events, int k) {
        int n = events.size();
        auto cmp = [&](const vector<int>& a, const vector<int>& b) {
            return a[1] < b[1];
        };
        ranges::sort(events, cmp);
        vector dp(n+1, vector<ll>(k+1));
        for(int i = 0; i < n; i++) {
            int start = events[i][0], end = events[i][1], val = events[i][2];
            int t = lower_bound(events.begin(), events.begin() + i, start, [](const vector<int>& e, int s) {
                return e[1] < s;
            }) - events.begin();
            for(int j = 0; j < k; j++) {
                dp[i+1][j+1] = max(dp[i][j+1], dp[t][j] + val);
            }
        }
        return dp[n][k];
    }
};

File: Dynamic_Programming/1751. Maximum Number of Events That Can Be Attended II.py
# TC:O(n * k + n * lgn) SC:O(n * k)
class Solution:
    def maxValue(self, events: List[List[int]], k: int) -> int:
        '''
        這題在基礎上多加了只能選k天限制 但是兩個變數時間複雜度來說還是1e6範圍內
        '''
        events.sort(key=lambda x:x[1])
        n = len(events)
        f = [[0] * (k+1) for _ in range(n+1)]
        for i in range(n):
            start, end, val = events[i]
            t = bisect_left(events, start, hi=i, key=lambda e: e[1])
            for j in range(k): # 因為是至多k個 所以 j > i + 1 也要列入計算 也就是答案回傳f[n][k]時要考慮少於k狀態
                f[i+1][j+1] = max(f[i][j+1], f[t][j] + val)
        return f[n][k]
